File: test/jsdom/networktest.test.mjs
// The in-product network self-test (/pending 23).
//
// **The discovery counters had a reader and it was a terminal.** `discovery.Socket.Stats()` counts
// sent, heard-ourselves and heard-peers at the same moments, and `nib discover` has printed a
// verdict off them since v1.117.18 — but Nib's primary user is non-technical, on one machine, with
// no IT, and a LAN ceremony that fails is silent by nature: a firewall, a VPN swallowing the group,
// an interface with no carrier.
//
// **D19 cannot cover this and that is structural, not an omission.** `diagnose()` returns
// `causeUndiagnosed` for a LAN or TCP ceremony by construction (`c.rz == nil || c.gate == nil`),
// and the status path publishes nothing for an undiagnosed cause — so the armed screen has never
// had a sentence for exactly this failure. This is that sentence.
//
// **What is driven here**: that the answer is the SERVER's sentence rather than a word rebuilt in
// the client, that the evidence beneath it is rendered, and that "nobody else is here" is not
// styled as success — it is an ordinary answer and still the reason nothing is happening.
//
// One boot per file — see boot.mjs.
import test from 'node:test';
import assert from 'node:assert/strict';
import { boot } from './boot.mjs';

let answer = {
  verdict: 'not-heard-back',
  summary: 'Nib sent announcements and did not hear its own come back. Something on this machine ' +
    'is blocking local network discovery — usually a firewall or a VPN. The other party will not ' +
    'hear you either.',
  sent: 12, own: 0, peers: 0, interfaces: 2, windowMs: 3000,
};
let asked = 0;

const { document: doc, settle } = await boot({
  routes: {
    '/api/lan/test': () => { asked += 1; return answer; },
  },
});

async function runTest() {
  doc.getElementById('srvNetTestGo').click();
  await settle();
  return doc.getElementById('srvNetTestOut');
}

test('the network test says what is wrong, in the words the server chose', async () => {
  const before = asked;
  const out = await runTest();

  // SETUP: the route was actually asked. Without this every assertion below is satisfied by a
  // control that renders a canned sentence and never calls anything.
  assert.equal(asked, before + 1, 'the control did not call /api/lan/test');

  const text = out.textContent;
  assert.ok(text.includes('blocking local network discovery'),
    "the server's own sentence is not rendered. The rule that decides which of the four states " +
    'obtains has ONE door and the wording travels with it — a client that rebuilt a sentence from ' +
    'the verdict tag would be a second copy of that rule, drifting from the CLI the day either ' +
    'is edited.');

  // The evidence beneath the sentence: the window separates "nobody is there" from "nobody
  // answered in three seconds", and the counters are what a user can quote.
  assert.match(text, /Listened for 3 seconds/,
    'the window is not rendered. A verdict with no window cannot be read: "nobody is there" and ' +
    '"nobody answered in three seconds" are different facts with the same sentence.');
  assert.match(text, /sent 12, heard itself 0, heard others 0/,
    'the counters the verdict was drawn from are not shown, so the answer can only be trusted ' +
    'and never checked — and a user reporting a problem has nothing to quote.');
  assert.match(text, /2 network connections/,
    'the interface count is dropped, and zero of them is its own failure — the one that arrives ' +
    'with no counters at all.');

  assert.ok(out.classList.contains('nettest-bad'),
    'a network that is dropping this machine\'s own multicast is not styled as a warning');
});

test('nobody else here is an ordinary answer, and still not success', async () => {
  answer = {
    verdict: 'nobody-else',
    summary: "This machine's network is working — Nib heard itself — and no other Nib is " +
      'announcing here. Either nobody else has armed a session yet, or they are on a different ' +
      'network.',
    sent: 6, own: 6, peers: 0, interfaces: 1, windowMs: 3000,
  };
  const out = await runTest();
  const text = out.textContent;

  assert.ok(text.includes('no other Nib is announcing here'),
    'the "nobody else" sentence is not rendered');
  assert.match(text, /1 network connection\b/,
    'the singular is wrong for one interface — this sentence is read by somebody already ' +
    'confused about their network');
  assert.ok(out.classList.contains('nettest-bad'),
    '"nobody else is here" is styled as success. The network works and the ceremony still is not ' +
    'starting, which is the whole reason the user pressed this — only a verdict of `working` is ' +
    'an answer that needs no action.');
});

test('a working network is the one answer that is not a warning', async () => {
  answer = {
    verdict: 'working', summary: 'Working: another Nib was heard on this network.',
    sent: 6, own: 6, peers: 3, interfaces: 1, windowMs: 3000,
  };
  const out = await runTest();
  assert.ok(out.textContent.includes('another Nib was heard'), 'the working sentence is missing');
  assert.equal(out.classList.contains('nettest-bad'), false,
    'a working network is styled as a warning, which would send a user hunting for a fault that ' +
    'is not there');
});

test('a test that could not run at all says why, and does not claim a clean network', async () => {
  answer = {
    verdict: 'nothing-sent',
    summary: 'Nib could not send anything on this network. No network connection on this machine ' +
      'accepted an announcement, so the other party cannot possibly hear you.',
    note: 'Nib could not listen on this network at all: no usable interface.',
    sent: 0, own: 0, peers: 0, interfaces: 0, windowMs: 3000,
  };
  const out = await runTest();
  const text = out.textContent;
  assert.ok(text.includes('could not send anything'), 'the verdict sentence is missing');
  assert.ok(text.includes('no usable interface'),
    'the note is dropped. It carries the reason the test could not run at all, and without it a ' +
    'machine with no network reads identically to one whose peers are simply absent.');
  assert.ok(out.classList.contains('nettest-bad'), 'a machine that can send nothing is not a warning');
});

File: internal/server/networktest_test.go
package server

import (
	"encoding/json"
	"net/http"
	"testing"

	"nib/internal/discovery"
)

// /pending 23 — the discovery counters get a surface that is not a terminal.
//
// **The counters had a reader and it was `nib discover`.** Nib's primary user is non-technical,
// on one machine, with no IT, and a LAN ceremony that fails is silent by nature. D19 cannot cover
// it either: `diagnose()` returns `causeUndiagnosed` for a LAN or TCP ceremony by construction,
// and the status path publishes nothing for an undiagnosed cause — so the armed screen has no
// sentence at all for this failure.
func TestTheNetworkTestAnswersWithAVerdictAndItsCounters(t *testing.T) {
	ts, _ := startServer(t)
	c, _ := authedClient(t, ts)

	r, err := c.Get(ts.URL + "/api/lan/test")
	if err != nil {
		t.Fatal(err)
	}
	defer r.Body.Close()
	if r.StatusCode != http.StatusOK {
		t.Fatalf("status = %d, want 200 — a diagnostic that faults is not a diagnostic, and a "+
			"machine with no usable interface is the case this route is most often asked about",
			r.StatusCode)
	}
	var got networkTestResponse
	if err := json.NewDecoder(r.Body).Decode(&got); err != nil {
		t.Fatal(err)
	}

	// The verdict is one of the four the door defines, and it is NAMED. An empty string would
	// leave a client branching on nothing.
	switch got.Verdict {
	case "working", "nobody-else", "not-heard-back", "nothing-sent":
	default:
		t.Errorf("verdict = %q, which is none of the four `discovery.Verdict` defines. A client "+
			"branches on this string, so an unknown value is a surface with no branch", got.Verdict)
	}
	if got.Summary == "" {
		t.Error("the network test returns no sentence. The whole item is that a non-technical " +
			"user cannot be sent to a terminal, and a machine tag with no prose is a terminal " +
			"answer wearing a JSON hat")
	}
	// **The window is reported, and it is the test's OWN window rather than the browse's.** A
	// reader has to tell "nobody is there" from "nobody answered in two seconds", and this route
	// deliberately listens longer than a browse because too few announce ticks produce a
	// confident, wrong "something is blocking discovery" on a healthy machine.
	if got.WindowMs != int(networkTestWindow.Milliseconds()) {
		t.Errorf("windowMs = %d, want %d — a verdict without the window it was measured over "+
			"cannot be read, and this route must not borrow the browse's shorter one",
			got.WindowMs, networkTestWindow.Milliseconds())
	}

	// The counters are published so the answer can be CHECKED rather than trusted, and they must
	// agree with the verdict — a summary that says one thing while its own evidence says another
	// is worse than no summary.
	st := discovery.Stats{Sent: got.Sent, Own: got.Own, Peers: got.Peers}
	if st.Verdict().Name() != got.Verdict {
		t.Errorf("the route reports verdict %q and counters (sent=%d own=%d peers=%d) that "+
			"classify as %q. The published counters are what let a user quote the evidence, so "+
			"they must be the evidence the verdict was drawn from",
			got.Verdict, got.Sent, got.Own, got.Peers, st.Verdict().Name())
	}
}

// The classifier's own truth table, driven directly — the route above can only ever produce
// whatever this machine's network happens to do, so the four states get asserted here.
//
// **`Sent` is checked FIRST and that ordering is the rule, not an implementation detail.** A
// machine that sent nothing also heard nothing of its own, so `Own == 0` is true there too;
// classifying on `Own` first would tell a user with no working interface that a firewall is
// dropping their multicast, which sends them to the wrong place entirely.
func TestTheVerdictSeparatesTheThreeWaysOfFindingNothing(t *testing.T) {
	for _, tc := range []struct {
		name string
		in   discovery.Stats
		want discovery.Verdict
	}{
		{"nothing left the machine", discovery.Stats{}, discovery.VerdictNothingSent},
		{"sent, and not one came back", discovery.Stats{Sent: 6}, discovery.VerdictNotHeardBack},
		{"heard ourselves, nobody else", discovery.Stats{Sent: 6, Own: 6}, discovery.VerdictNobodyElse},
		{"a peer answered", discovery.Stats{Sent: 6, Own: 6, Peers: 1}, discovery.VerdictWorking},
		// The ordering case: nothing was sent AND nothing came back. Only one of those is the
		// user's actual problem.
		{"neither sent nor heard", discovery.Stats{Own: 0, Peers: 0}, discovery.VerdictNothingSent},
	} {
		t.Run(tc.name, func(t *testing.T) {
			if got := tc.in.Verdict(); got != tc.want {
				t.Errorf("Verdict() = %v, want %v for %+v", got, tc.want, tc.in)
			}
			if tc.in.Verdict().Summary() == "" {
				t.Error("this verdict has no sentence for a person")
			}
			if tc.in.Verdict().Name() == "" {
				t.Error("this verdict has no machine tag")
			}
		})
	}

	// Every verdict says something DIFFERENT. Four states folded into one sentence is the defect
	// this whole item is about, one level up.
	seen := map[string]string{}
	for _, v := range []discovery.Verdict{
		discovery.VerdictWorking, discovery.VerdictNothingSent,
		discovery.VerdictNotHeardBack, discovery.VerdictNobodyElse,
	} {
		if prev, dup := seen[v.Summary()]; dup {
			t.Errorf("%q and %q share a sentence, so a user cannot tell them apart", prev, v.Name())
		}
		seen[v.Summary()] = v.Name()
	}
	if len(seen) != 4 {
		t.Errorf("four verdicts produced %d distinct sentences", len(seen))
	}
}
